internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,894 @@ var _ service.ChunkService = &ChunkServiceHandler{}
+}
+
+type mvInfo struct {
+	mu           sync.Mutex
+	mvName       string   // mv0, mv1, etc.
+	componentRVs []string // sorted list of component RVs for this MV
+
+	// count of chunk operations (get, put or remove) for this MV.
+	// This is used to block the end sync call till all the ongoing chunk operations are completed.
+	chunkOpsCnt atomic.Int64
+
+	// flag to block chunk operations (get, put or remove) for this MV.
+	// This flag is enabled in the end sync call to pause new chunk operations in the MV.
+	// When the end sync call is completed, this flag is disabled.
+	blockOpsFlag atomic.Bool
+
+	syncInfo // sync info for this MV
+}
+
+type syncInfo struct {
+	// Is the MV in syncing state?
+	// An MV enters syncing state after StartSync command is successfully executed.
+	// In syncing state, PutChunk requests corresponding to client writes will be
+	// saved in the mv#.sync folder. Similarly a successful EndSync will take an
+	// MV out of syncing state.
+	isSyncing bool
+
+	// sync ID for this MV if it is in syncing state.
+	// This is returned in the StartSync response and EndSync should carry this.
+	syncID string
+
+	// sourceRV is the RV that is syncing the target RV in the MV.
+	// This is used when the source RV goes offline and is replaced by the cluster manager with a new RV.
+	// In this case, the sync will need to be restarted from the new RV.
+	sourceRV string // source RV for syncing this MV
 }
 
 func NewChunkServiceHandler() *ChunkServiceHandler {
+	// TODO:: integration: get rvID, rvName and cache dir path for different RVs for the node from cluster manager
+	// below will be call to cluster manager to get the information
+	rvIDMap := getRvIDMap()
+
 	return &ChunkServiceHandler{
-		locks: common.NewLockMap(),
+		locks:   common.NewLockMap(),
+		rvIDMap: rvIDMap,
+	}
+}
+
+// check if the given mv is valid
+func (rv *rvInfo) isMvValid(mvPath string) error {
+	if !common.DirectoryExists(mvPath) {
+		return fmt.Errorf("MV path %s does not exist", mvPath)
+	}
+
+	mvName := filepath.Base(mvPath)
+	_, err := rv.getMVInfo(mvName)
+	return err
+}
+
+func (rv *rvInfo) getComponentRVsForMV(mvName string) []string {
+	mvInfo, err := rv.getMVInfo(mvName)
+	if err != nil {
+		log.Warn("ChunkServiceHandler::getComponentRVsForMV: Failed to get MV %s info for %s [%v]", mvName, rv.rvName, err.Error())
+		return nil
+	}
+
+	return mvInfo.componentRVs
+}
+
+// getMVInfo returns the mvInfo for the given mvName
+func (rv *rvInfo) getMVInfo(mvName string) (*mvInfo, error) {
+	val, ok := rv.mvMap.Load(mvName)
+	mvInfo := val.(*mvInfo)
+	if !ok || mvInfo == nil {
+		common.Assert(false, "MV %s not found in RV %s", mvName, rv.rvName)
+		return nil, fmt.Errorf("MV %s is invalid for RV %s", mvName, rv.rvName)
+	}
+
+	common.Assert(mvName == mvInfo.mvName, "MV name mismatch in mv", mvName, mvInfo.mvName)
+
+	return mvInfo, nil
+}
+
+func (rv *rvInfo) addToMVMap(mvName string, val *mvInfo) {
+	rv.mvMap.Store(mvName, val)
```

Comment:
we can add a comment that caller is expected to ensure that the mv path does exist, since we later rely on that.

Reply:
Also, since we should only call addToMVMap() on receiving a JoinMV request and join-mv must be called only by cluster manager when it's adding a new rv to an mv and never add the same mv again, we can assert that the mv is not already present. In prod we should fail such a request since clustermap is definitely not inline with our understanding of membership.

Reply:
The caller of this method is `JoinMV` which already checks if the RV is part of the given MV, before calling this method. So, we don't need to check this again.
Also, added comment that the caller of this method must ensure that the RV is not part of the given MV.

Reply:
Added assert